src/imp/linux_raw/syscalls.rs

```diff
@@ -3028,17 +3028,16 @@ pub(crate) fn ttyname(fd: BorrowedFd<'_>, buf: &mut [u8]) -> io::Result<()> {
     // Gatter the ttyname by reading the 'fd' file inside 'proc_self_fd'
     let r = readlinkat(proc_self_fd, DecInt::from_fd(&fd).as_c_str(), buf)?;
 
-    // Make if possible the string nul terminated
-    if r == buf.len() {
+    // If the number of bytes is equal to the buffer length, truncation may
+    // have occurred. And we need one extra byte for a NUL terminator.
+    if buf.len() - r <= 1 {
```

Comment:
But I don't understand why you changed that.

Reply:
[musl](https://github.com/ifduyue/musl/blob/3f701faace7addc75d16dea8a6cd769fa5b3f260/src/unistd/ttyname_r.c#L18) had the same exact check.

Reply:
Ah, right. I was thinking we needed to reserve an extra byte for the NUL terminator, but the original check effectively does that. I've restored the original check now.